src/hooks.ts

```diff
@@ -151,46 +202,56 @@ const initializeWhenClientReadyFn = (
  *       ClientReady and DidTimeout provide signals to handle this scenario.
  */
 export const useExperiment: UseExperiment = (experimentKey, options = {}, overrides = {}) => {
-  const { isServerSide, optimizely, timeout } = useContext(OptimizelyContext);
+  const { optimizely, isServerSide } = useContext(OptimizelyContext);
   if (!optimizely) {
     throw new Error('optimizely prop must be supplied via a parent <OptimizelyProvider>');
   }
 
-  // Helper function to return the current value for variation.
-  const getCurrentValues = useCallback<() => ExperimentDecisionValues>(
+  const overrideAttrs = useCompareAttrsMemoize(overrides.overrideAttributes);
+  const getCurrentDecision: () => ExperimentDecisionValues = useCallback(
     () => ({
-      variation: optimizely.activate(experimentKey, overrides.overrideUserId, overrides.overrideAttributes),
+      variation: optimizely.activate(experimentKey, overrides.overrideUserId, overrideAttrs),
     }),
-    [experimentKey, overrides]
+    [optimizely, experimentKey, overrides.overrideUserId, overrideAttrs]
   );
 
-  // Set the initial state immediately serverSide
-  const [state, setState] = useState<UseExperimentState>(() => {
-    const initialState = {
-      variation: null,
-      clientReady: isServerSide ? true : false,
-      didTimeout: false,
-    };
-    if (isServerSide) {
-      return { ...initialState, ...getCurrentValues() };
+  const isClientReady = isServerSide || optimizely.isReady();
+  const decisionStateAndSetter = useState<ExperimentDecisionValues>(() => {
+    if (isClientReady) {
+      return getCurrentDecision();
     }
-    return initialState;
+    return { variation: null };
   });
+  let decisionState = decisionStateAndSetter[0];
+  const setDecisionState = decisionStateAndSetter[1];
+  // Decision state is derived from entityKey and overrides arguments.
+  // Track the previous value of those arguments, and update state when they change.
+  // This is an instance of the derived state pattern recommended here:
+  // https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops
+  const currentDecisionInputs: DecisionInputs = {
+    entityKey: experimentKey,
+    overrideUserId: overrides.overrideUserId,
+    overrideAttributes: overrideAttrs,
+  };
+  const [prevDecisionInputs, setPrevDecisionInputs] = useState<DecisionInputs>(currentDecisionInputs);
+  if (!areDecisionInputsEqual(prevDecisionInputs, currentDecisionInputs)) {
+    setPrevDecisionInputs(currentDecisionInputs);
+    decisionState = getCurrentDecision();
+    setDecisionState(decisionState);
```

Comment:
I might be missing something obvious but I can't see what `setDecisionState` does with `decisionState`. It looks like a function without any params.

Reply:
`setDecisionState` is a function, it's the 2nd argument returned from `useState` on line 219. I am trying to update the state value here.